### DISCUSSION TITLE
v0.20 release

### DISCUSSION THREAD

**Discussion by ssheorey** [isl-org/Open3D#7564] [General]
# Open3D v0.20

Open3D v0.20 brings Gaussian Splatting to the Open3D renderer, makes Intel GPU workflows substantially more practical through expanded SYCL Tensor pipelines, and improves the developer experience with smaller installs, Python type checking, and new distribution options.

### Gaussian Splatting

<video width="640" height="480" controls>
  <source src="https://github.com/user-attachments/assets/f2840a2a-ffea-4347-9039-e1ecfd089fb2" type="video/mp4">
  Open3D Gaussian Splatting demo.
</video>

https://github.com/user-attachments/assets/f2840a2a-ffea-4347-9039-e1ecfd089fb2

*Gaussian splats and meshes in one Open3D scene.*

```sh
python examples/python/visualization/draw_from_csv.py
```

Open3D can now render 3D Gaussian Splats in interactive and offscreen views, along with regular PBR meshes.  Load tensor point clouds from Gaussian PLY, binary SPLAT, or compressed SPZ files, then compose them with triangle meshes using correct depth occlusion.  This enables applications to combine radiance-field captures with conventional 3D assets in the same Open3D scene (PRs [#7191](https://github.com/isl-org/Open3D/pull/7191), [#7194](https://github.com/isl-org/Open3D/pull/7194), [#7490](https://github.com/isl-org/Open3D/pull/7490), [#7520](https://github.com/isl-org/Open3D/pull/7520), and [#7525](https://github.com/isl-org/Open3D/pull/7525); contributed by @lumurillo, @billamiable, @ssheorey, and @jeevan6996).

The renderer supports anti-aliasing, depth output and offscreen rendering (see `OffscreenRenderer`). See the [Gaussian Splatting tutorial](jupyter/visualization/gaussian_splatting.ipynb), the mixed-scene [Python example](../examples/python/visualization/draw_from_csv.py), and the [C++ example](../examples/cpp/GaussianSplat.cpp).

<img width="480" alt="Horned Lizard annotation" src="https://github.com/user-attachments/assets/59854e0a-8ed6-4515-8112-10aff12e15da" />

*Labeling geometry over a compressed SPZ Gaussian Splat.*

### Intel GPU Pipelines Through SYCL

<video width="640" height="480" controls>
  <source src="https://github.com/user-attachments/assets/44f03300-c887-4839-9855-1fbd50e15ac0" type="video/mp4">
  RGBD dense SLAM on Intel B580.
</video>

https://github.com/user-attachments/assets/44f03300-c887-4839-9855-1fbd50e15ac0

*Real time RGBD reconstruction with dense SLAM on an Intel B580 GPU*

```sh
pip install open3d-xpu
cd examples/python/t_reconstruction_system
python dense_slam.py --device SYCL:0 --default_dataset lounge --path_npz lounge_sycl.npz
```

The SYCL backend now supports substantially more Tensor-based Open3D pipelines on Intel integrated and discrete GPUs. New native-device implementations cover HashMap, nearest-neighbor search, geometry transforms, registration, RGB-D odometry, feature computation, TSDF reconstruction, and raycasting. Hybrid nearest-neighbor search now supports L1 and Linf metrics in addition to L2 (PRs [#7114](https://github.com/isl-org/Open3D/pull/7114) and [#7443](https://github.com/isl-org/Open3D/pull/7443); contributed by @ssheorey).

The newly implemented Intel GPU algorithms include:

- **Core and geometry:** device HashMap; KNN, fixed-radius, and hybrid nearest-neighbor search; Tensor geometry and normal transforms; TriangleMesh normal and area computation; and partial Tensor PointCloud unprojection, normal, and NNS-based operations.
- **Registration and odometry:** point-to-point, point-to-plane, colored, Doppler, and symmetric ICP; registration evaluation and information matrices; FPFH features and correspondence matching; and point-to-plane, intensity, and hybrid RGB-D odometry.
- **Reconstruction and raycasting:** VoxelBlockGrid TSDF touch, integration, ray casting, point-cloud extraction, and marching-cubes mesh extraction, together with hardware-accelerated `RaycastingScene` queries.

Open3D-ML operations are also available through the PyTorch XPU library, including continuous and sparse convolutions, voxelization and pooling, neighbor-list operations, NMS, RoI pool, point queries, IoU, and interpolation.

Prebuilt SYCL C++ archives and `open3d-xpu` wheels are available for Ubuntu 22.04+ and Windows 10+. The Windows `open3d-xpu` wheel is a preview release; it and the Windows `open3d-cuda` wheel should be treated as experimental while we expand platform coverage. See the [SYCL documentation](sycl.rst) for device requirements, installation, and the supported Tensor API matrix.

### Smaller Wheels and Better Python Tooling And Wider cross-platform support

A new install of the `open3d` wheel is now approximately **335 MB smaller**, through more efficient C++ library packing and moving ML dependencies to `extras` (PRs [#7516](https://github.com/isl-org/Open3D/pull/7516) and [#7540](https://github.com/isl-org/Open3D/pull/7540); contributed by @ssheorey).  Open3D-ML Python dependencies now install only when specifically requested:

```shell
pip install "open3d[ml]"
```

This only affects python dependencies - the full core functionality is included without `[ml]`.  

Open3D now ships Python type information for static type checking and richer editor autocomplete. This improves development with tools such as mypy and VS Code without changing the dynamic Python API.

<img width="512" alt="Python type checks and inline docs." src="https://github.com/user-attachments/assets/f460cb22-e493-4122-9013-cd12c2881f21" />

Open3D supports Python 3.10 through 3.14, including new Python 3.13 and 3.14 wheels. PyTorch is updated to 2.13 and TensorFlow to 2.20. TensorFlow operations are not available on Python 3.14, and compiled TensorFlow operations remain unsupported on Windows. Open3D validates the installed framework's major and minor version against the version used to build its ML operations.

Open3D now has wheels for more platforms:

```sh
# Linux x86_64 CUDA wheel and Windows, macOS, Linux aarch64 CPU wheel:
pip install open3d
# Linux and Windows XPU wheel (Intel integrated and discrete GPUs)
pip install open3d-xpu
# Windows x86_64 CUDA preview wheel
pip install https://github.com/isl-org/Open3D/releases/download/v0.20.0/open3d-0.20.0-cp312-cp312-win_amd64.whl  # Python 3.10-3.14
```

### Registration, Geometry, and Reconstruction

- Added symmetric ICP registration for legacy and Tensor pipelines (PR [#7276](https://github.com/isl-org/Open3D/pull/7276); contributed by @eclipse0922).
- Added 3D Normal Distributions Transform (NDT) registration with C++ and Python APIs (PR [#7517](https://github.com/isl-org/Open3D/pull/7517); contributed by @Xiang-Zeng).
- Added Moving Least Squares, Laplacian, Taubin, and bilateral point-cloud smoothing (PR [#7419](https://github.com/isl-org/Open3D/pull/7419); contributed by @TsofnatMaman).
- Exposed `full_depth`, `samples_per_node`, and `point_weight` for Poisson surface reconstruction (PR [#7430](https://github.com/isl-org/Open3D/pull/7430); contributed by @Chevi-Koren).
- Added `CorrespondenceCheckerBasedOnSourceRotation` for RANSAC workflows with a known global orientation prior (PR [#7461](https://github.com/isl-org/Open3D/pull/7461); contributed by @jGiltinan).
- Added accurate minimum-volume oriented bounding boxes using Jylanki's algorithm `OrientedBoundingBox::CreateFromPointsMinimal()`, alongside the fast approximate method (PR [#7138](https://github.com/isl-org/Open3D/pull/7138); contributed by @quentin-leboutet).

  <table>
      <tr>
      <td><img src="https://github.com/user-attachments/assets/5b1e7b5a-9804-4b21-a4c3-1db845aba5d8" width="420"></td>
      <td><img src="https://github.com/user-attachments/assets/12200062-7c82-40af-8a8c-ba0dd6181d6b" width="420"></td>
    </tr>
  </table>

- Added minimum-volume oriented bounding ellipsoids for point clouds and triangle meshes with Khachiyan's algorithm (PR [#7377](https://github.com/isl-org/Open3D/pull/7377); contributed by @quentin-leboutet).

  <img width="480" alt="Minimum volume oriented bounding ellipsoid." src="https://github.com/user-attachments/assets/b41d3137-5b38-4f89-bad3-a639154e7eec" />

- Added `transform`, `translate`, `scale`, and `rotate` support for legacy `VoxelGrid` geometry (PR [#7339](https://github.com/isl-org/Open3D/pull/7339); contributed by @lmuffang).

### Rendering, Interoperability, and Remote Visualization

- Added MikkTSpace tangent space generation, normal-map conversion, and ray-cast ambient-occlusion texture baking for Tensor triangle meshes (PR [#7364](https://github.com/isl-org/Open3D/pull/7364); contributed by @ssheorey).

  <table>
    <tr>
      <th>Baked tangents</th>
      <th>Baked normals</th>
    </tr>
    <tr>
      <td><img alt="Baked tangents" src="https://github.com/user-attachments/assets/77f72a66-194b-49ad-90c6-851de95d855f" width="256"></td>
      <td><img alt="Baked normals" src="https://github.com/user-attachments/assets/76b4d13b-7eed-45f5-ba65-6dd5457eade2" width="256"></td>
    </tr>
    <tr>
      <th>Original world-space normal map</th>
      <th>Computed tangent-space normal map</th>
    </tr>
    <tr>
      <td><img alt="Original world-space normal map" src="https://github.com/user-attachments/assets/5622b34a-b343-4bd1-9fb6-88287901be56" width="256"></td>
      <td><img alt="Computed tangent-space normal map" src="https://github.com/user-attachments/assets/cc04fcb1-ec71-44cf-8041-eef53f24a793" width="256"></td>
    </tr>
    <tr>
      <th>Restored world-space normal map</th>
      <th>Ambient occlusion for a torus</th>
    </tr>
    <tr>
      <td><img alt="Restored world-space normal map" src="https://github.com/user-attachments/assets/583af190-54e3-4aa2-9fc7-4501da2fab12" width="256"></td>
      <td><img alt="Ambient occlusion for a torus" src="https://github.com/user-attachments/assets/2c9386e3-2268-4e22-bb8f-f0dac35f4aca" width="256"></td>
    </tr>
  </table>
- Added experimental USD/USDZ import and transparent Draco decompression for lights (PR [#7505](https://github.com/isl-org/Open3D/pull/7505); contributed by @ssheorey); broader scene semantics are not yet supported.

  ![USD toy biplane](https://github.com/user-attachments/assets/ab4ddbc3-0f5b-4caf-9ddb-3c8be88bc02f)

- Reduced WebRTC remote-visualization delays with input/redraw coalescing, asynchronous frame encoding, and streaming configuration improvements (PR [#7496](https://github.com/isl-org/Open3D/pull/7496); contributed by @ssheorey).
- Replaced OSMesa with EGL for Linux legacy-visualizer headless rendering (PR [#7518](https://github.com/isl-org/Open3D/pull/7518); contributed by @ssheorey). Standard Open3D Linux binaries can render without `DISPLAY` or `WAYLAND_DISPLAY` when an EGL-capable driver is available.

### Distribution and Build Experience

- All Open3D distributed binaries now have [GitHub artifact attestations](https://docs.github.com/en/actions/concepts/security/artifact-attestations) (PR [#7512](https://github.com/isl-org/Open3D/pull/7512)). Users can verify that the distribution they are using was built on Open3D GitHub CI.

```sh
gh attestation verify /path/to/<artifact> -R isl-org/Open3D
```

- Open3D now uses oneAPI TBB for CPU parallelism, removing its OpenMP runtime dependency and avoiding common runtime conflicts with frameworks such as PyTorch (PR [#6626](https://github.com/isl-org/Open3D/pull/6626); contributed by @dbs4261).
- Added vcpkg support for dependency management (PR [#7386](https://github.com/isl-org/Open3D/pull/7386); contributed by @eclipse0922).
- The Open3D Viewer is available through the Microsoft Store on Windows, with file associations for common geometry formats including `.splat` and `.spz` (PR [#7534](https://github.com/isl-org/Open3D/pull/7534); contributed by @ssheorey).
- Repaired WebRTC prebuilt packaging across Linux, macOS arm64, and Windows runtime variants (PR [#7515](https://github.com/isl-org/Open3D/pull/7515); contributed by @ssheorey).

### Agent Skills for Open3D Applications

Open3D now ships agent skills for AI coding assistants that write applications using Open3D (PR [#7542](https://github.com/isl-org/Open3D/pull/7542); contributed by @eclipse0922). The `open3d-python` and `open3d-cpp` skills map the public APIs, clarify the Tensor-versus-legacy API choices, and cover device, I/O, visualization, registration, reconstruction, and CMake integration.

The skills are included in Python wheels and C++ binary packages. Install the Python skill into a project's `.github/skills` directory with:

```shell
open3d agent_skill --install .github/skills
```

For C++, copy `open3d-cpp` from `share/Open3D/agent_skills/open3d-cpp` in an extracted `open3d-devel` package to the project's `.github/skills/` directory. On Windows, the source directory is `bin/Open3D/agent_skills/open3d-cpp`.

### Correctness and Performance

- Non-contiguous NumPy and Pandas vector conversion is up to about 50x faster in affected workloads (PR [#7343](https://github.com/isl-org/Open3D/pull/7343); contributed by @Khizar1).
- Mesh point sampling now samples mesh colors when present and correctly uses area-uniform triangle sampling (PRs [#6842](https://github.com/isl-org/Open3D/pull/6842) and [#6653](https://github.com/isl-org/Open3D/pull/6653); contributed by @nikste and @ghost).
- CUDA now handles sliced boolean-mask indexing and large KNN selection more reliably (PRs [#7340](https://github.com/isl-org/Open3D/pull/7340) and [#7381](https://github.com/isl-org/Open3D/pull/7381); contributed by @TwentyPast4 and @nicolaloi).

### CPU Parallelism Uses TBB

OpenMP support has been removed. `WITH_OPENMP`, `OMP_NUM_THREADS`, `utility::OMPProgressBar`, `utility::GetThreadNum()`, and `utility::InParallel()` are no longer available. Use `open3d.utility.set_max_threads()` in Python or `utility::SetMaxThreads()` in C++; applications that require stricter thread control can use a TBB task arena. Open3D requires oneTBB 2021.4.0 or newer.

### ML Dependencies Are Optional

`pip install open3d` no longer installs Open3D-ML Python dependencies. Install the `ml` extra when importing Open3D-ML functionality:

```shell
pip install "open3d[ml]"
```

### Platform and Package Compatibility

- macOS x86_64 is no longer supported; macOS distributions are arm64-only.
- Python package DLL discovery no longer searches `PATH` for non-CUDA DLLs.
- Windows `open3d-cuda` and `open3d-xpu` (sycl) wheels and devel packages are preview/experimental. Windows `cuda` does not support PyTorch ops, and TensorFlow ops is not available on Windows. Debug devel builds for `cuda` and `sycl` are not available.
- Linux legacy-visualizer headless builds now use EGL instead of OSMesa. The `ENABLE_HEADLESS_RENDERING` and `HEADLESS_RENDERING` CMake options and OSMesa dependencies have been removed. Windows and macOS still require an active desktop/window-server session for legacy-visualizer rendering.

### SYCL Limitations

As with CUDA, some legacy APIs are not yet available. Open3D currently uses one SYCL GPU at a time. GPU kernels are JIT-compiled on first use, so the first run on a device can be slower; enable a persistent cache with `o3d.core.sycl.enable_persistent_jit_cache()` or `SYCL_CACHE_PERSISTENT=1`.

### Gaussian Splatting Limitations

Some platforms may hang or crash when displaying Gaussian Splats (Windows Intel Arc A770m is known to hang.). Gaussian Splatting uses Vulkan on Linux and Windows and Metal on macOS; the OpenGL renderer is not supported. MSAA is disabled for splat views, with the Gaussian anti-aliasing option serving that role. Very large or close splats can exceed tile limits and be culled, and dense scenes can drop tile entries after reaching their configured capacity. Use the Gaussian Splat material capacity parameters to tune these limits for the scene.

For anisotropic splats, do not use a general 4x4 `Transform()` to change the scene. Use `Translate`, `Scale`, and `Rotate` so positions, covariance orientation, and view-dependent coefficients remain consistent.

## Acknowledgment

Thank you to the Open3D community contributors, issue reporters, and users who helped shape this release.

## Full Changelog

[v0.19.0...v0.20.0](https://github.com/isl-org/Open3D/compare/v0.19.0...v0.20.0)

<hr /><em>This discussion was created from the release <a href='https://github.com/isl-org/Open3D/releases/tag/v0.20.0'>v0.20 release</a>.</em>
